tests/test_merge.py

```diff
@@ -379,6 +381,332 @@ def test_merge(
+        ),
+        param(
+            (DictConfig(content={}, element_type=Optional[str]), {"foo": "qux"}),
+            {"foo": "qux"},
+            str,
+            True,
+            id="new_optional_str",
+        ),
+        param(
+            (DictConfig(content={}, element_type=Optional[str]), {"foo": None}),
+            {"foo": None},
+            str,
+            True,
+            id="new_optional_str_none",
+        ),
+        param(
+            (DictConfig(content={}, element_type=Optional[str]), {"foo": MISSING}),
+            {"foo": MISSING},
+            str,
+            True,
+            id="new_optional_str_missing",
+        ),
+        param(
+            (DictConfig(content={"foo": MISSING}, element_type=str), {"foo": "qux"}),
+            {"foo": "qux"},
+            str,
+            False,
+            id="missing_str",
+        ),
+        param(
+            (DictConfig(content={"foo": MISSING}, element_type=str), {"foo": None}),
+            raises(ValidationError, match="Non optional field"),
+            None,
+            None,
+            id="missing_str_none",
+        ),
+        param(
+            (DictConfig(content={"foo": MISSING}, element_type=str), {"foo": MISSING}),
+            {"foo": MISSING},
+            str,
+            False,
+            id="missing_str_missing",
+        ),
+        param(
+            (
+                DictConfig(content={"foo": MISSING}, element_type=Optional[str]),
+                {"foo": "qux"},
+            ),
+            {"foo": "qux"},
+            str,
+            True,
+            id="missing_optional_str",
+        ),
+        param(
+            (
+                DictConfig(content={"foo": MISSING}, element_type=Optional[str]),
+                {"foo": None},
+            ),
+            {"foo": None},
+            str,
+            True,
+            id="missing_optional_str_none",
+        ),
+        param(
+            (
+                DictConfig(content={"foo": MISSING}, element_type=Optional[str]),
+                {"foo": MISSING},
+            ),
+            {"foo": MISSING},
+            str,
+            True,
+            id="missing_optional_str_missing",
+        ),
+        param(
+            (
+                DictConfig(content={"foo": User("Bond")}, element_type=User),
+                {"foo": User("Bond")},
+            ),
+            {"foo": User("Bond")},
+            User,
+            False,
+            id="user",
+        ),
+        param(
+            (
+                DictConfig(content={"foo": User("Bond")}, element_type=User),
+                {"foo": None},
+            ),
+            raises(ValidationError, match="field 'foo' is not Optional"),
+            None,
+            None,
+            id="user_none",
+        ),
+        param(
+            (
+                DictConfig(content={"foo": User("Bond")}, element_type=User),
+                {"foo": MISSING},
+            ),
+            {"foo": MISSING},
```

Comment:
Shouldn't this one be `{"foo": User("Bond")}`?

Reply:
Good catch! Fixed in https://github.com/omry/omegaconf/pull/749/commits/9a2293956d9a6663d305e11b57ce077e867d21de, and the test is passing now :)